### DISCUSSION TITLE
Related model uses parent connection when `$connection` is `null`

### DISCUSSION THREAD

**Discussion by lombervid** [laravel/framework#52735] [Ideas]
### Laravel Version

11.19.0

### PHP Version

8.3.8

### Database Driver & Version

MySQL 5.7.44

### Description

Having a model, without `$connection` property specified, as a relation from another model with explicit `$connection` causes the child to use the parent's connection instead of the default connection.

I found out this behavior was implemented in #16103 but I think it might not be the desired behavior.
According with what the documentation says, it should use the default connection (and that was exactly what I was expecting):

> By default, all Eloquent models will use the default database connection that is configured for your application. If you would like to specify a different connection that should be used when interacting with a particular model, you should define a `$connection` property on the model:

### Steps To Reproduce

Child Model:
```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Child extends Model
{
    //
}
```

Parent Model:
```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class Parent extends Model
{
    protected $connection = 'another';

    public function child(): BelongsTo
    {
        // `Child` model will use `another` connection instead of the default
        return $this->belongsTo(Child::class);
    }
}
```

### Proposed solution:

I think it could be implemented something like this:

Model.php
```php
<?php

namespace Illuminate\Database\Eloquent;

abstract class Model
{
    // ...
    protected $modelShouldInheritConnection = false;

    // ...

    public function inheritsConnection(): bool
    {
        return $this->$modelShouldInheritConnection;
    }

    // ...
}
```

HasRelationships.php
```php
<?php

namespace Illuminate\Database\Eloquent\Concerns;

trait HasRelationships
{
    // ...

    protected function newRelatedInstance($class)
    {
        return tap(new $class, function ($instance) {
            if ($instance->inheritsConnection() && ! $instance->getConnectionName()) {
                $instance->setConnection($this->connection);
            }
        });
    }

    // ...
}
```

Then in the model you want to inherit the parent's connection:
```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Child extends Model
{
    protected $modelShouldInheritConnection = true;
}
```

Or it could be set to `true` by default so as not to break the current behavior.

Kind regards!

**Comment by ghost** [comment]
Hi. From what I know the relations (by default) are considered to be in the same connection. So for me this is expected behavior.

**Comment by lombervid** [comment]
I would say it depends on the use case.

For example, in my case, I had many models (all using the default connection). Then I needed to add a model with a new connection, and, in that model to have a relation with another (already created, with default connection) model. So, my expected behavior was that the model I was already using with the default connection would keep using the same connection it was previously using.

But, either way, I think it would be nice to have a way to explicitly specify the desired behavior to use in a model. 

I think it could be a static property to be able to set it for all models as well. Something like:

`Model.php`
```php
<?php

namespace Illuminate\Database\Eloquent;

abstract class Model
{
    // ...
    protected static $modelShouldInheritConnection = false;

    // ...

    public static function inheritsConnection(): bool
    {
        return static::$modelShouldInheritConnection;
    }

    public static function shouldInheritsConnection(bool $value = true): void
    {
        static::$modelShouldInheritConnection = $value;
    }

    // ...
}
```

Then we could be able to set in the `AppServiceProvider` something like:
```php
Model::shouldInheritsConnection(true);
```

And also be able to change the behavior in a specific model like:
```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Child extends Model
{
    protected static $modelShouldInheritConnection = false;
}
```

**Reply by ghost** [reply]
Can't you just use the connection property that exists in the model?

**Reply by lombervid** [reply]
Yeah, that could be a solution. But then the property has to be set in every model you are using that way. And, in the remote case, if you change the default connection, you would have to change the value in every model you had specified the connection property.

**Reply by macropay-solutions** [reply]
@lombervid we are using a base abstract model for that https://github.com/macropay-solutions/laravel-crud-wizard-free/blob/production/src/Models/BaseModel.php
You could also use it like that or you could try using a trait.
